WAGE: New Engine

sev- opened the pull request:
Here is my PR for the WAGE engine, a WorldBuider (https://en.wikipedia.org/wiki/World_Builder) games. It was a nice and fun side project of me. There are over 100 games built on this engine.

I've tested few smaller games and they're completable, although a more thorough testing is required.

TODO:
- Sound (very simple SFX, sometimes looped, no music)
- About dialog
- Shortcuts from Actions menu
- Save/load
- Improvements in the input. It still doesn't match the original
- Copy/paste in the text console window
- Add more games to detection
- Window closing (I do not plan to implement window resizing like in original)
- General bugfixing

Also, there are some uncertainties and open questions in the game logic, primarily because there are not many games using those features. The whole engine is based on WAGE engine written in Java by Alex Svitkine (https://github.com/asvitkine/wage-engine) which in turn is mainly a guesswork. I plan to try to chase the original WorldBuilder author in a hope that he shares the sources with us.

In comparison to the Alex's engine, I strive to be pixel perfect and implement many features by referring to the original games running in the vMac emulator.

The engine optionally uses the original Mac fonts, there is a Mac-only script to downloads System 7.0.1 image from Apple (the OS has been put as freeware), extracts it and puts into wage.dat file. It is not clear to me whether we're allowed to redistribute the fonts. As a fallback, the engine uses built-in ScummVM fonts, and that is naturally not pixel-perfect.

The majority of these games were either freeware or shareware, so I plan to prepare 2 archives: originals and extracted games in the MacBinary format ready for ScummVM consumption.

bluegr commented:
Nice work, kudos! :)

I'll be able to check this out tomorrow

bluegr commented on devtools/create_wage/create_wage.sh:
```diff
@@ -0,0 +1,119 @@
+
+echo_n "Copying fonts..."
+
+for i in Athens Cairo Chicago Courier Geneva Helvetica London "Los Angeles" Monaco "New York" Palatino "San Francisco" Symbol Times Venice
```
Do we actually need all of these fonts? Cairo is the default and most probably necessary, but what about the others?

Can we use Liberation fonts as fallback, instead of the ScummVM built-in ones?
https://fedorahosted.org/liberation-fonts/

sev- replied:
Yes, fonts are referred in the game data. I saw at least 3 fonts used, and the game authors are free to use whatever they want.

As of the Liberation fonts, they will not be pixel-perfect, thus, in my opinion, not much better than the currently used built-in ones.

bluegr commented on engines/wage/combat.cpp:
```diff
@@ -0,0 +1,915 @@
+
+namespace Wage {
+
+Obj *WageEngine::getOffer() {
```
Um, what is an "offer" here? It's not really apparent. Could you please add a comment to explain what this does?

sev- replied:
Part of the gameplay. There is such command 'offer' when the hero may offer something to an NPC.

bluegr commented on engines/wage/randomhat.h:
```diff
@@ -0,0 +1,77 @@
+	kTokNone = -100000
+};
+
+class RandomHat {
```
I don't understand what this does... It seems to generate random values for weapon, magic etc usage based on tokens? Could you please add a short description?

sev- replied:
Draws random number from classes with probability. Another part of the gameplay.

bluegr commented on engines/wage/combat.cpp:
```diff
@@ -0,0 +1,915 @@
+}
+
+void WageEngine::encounter(Chr *player, Chr *chr) {
+	char buf[512];
```
Perhaps it would be better to use Common::String here, along with Common::String::format() instead of snprintf() below?

sev- replied:
Will do.

bluegr commented on engines/wage/combat.cpp:
```diff
@@ -0,0 +1,915 @@
+		for (ObjList::const_iterator it = objs->begin(); it != objs->end(); ++it, ++cnt) {
+			if ((*it)->_type != Obj::IMMOBILE_OBJECT) {
+				// TODO: I'm not sure what the chance should be here.
+				hat.addTokens(cnt, 123);
```
Does this actually happen? Should we use a warning here when this case occur?

sev- replied:
It does. Just since the engine based on guesswork, actual number of token is not known.

bluegr commented on engines/wage/combat.cpp:
```diff
@@ -0,0 +1,915 @@
+	switch (token) {
+		case kTokWeapons:
+			// TODO: I think the monster should choose the "best" weapon.
+			performAttack(npc, player, weapons->operator[](_rnd->getRandomNumber(weapons->size() - 1)));
```
Again, does this actually happen? Is there a test case? (same with kTokMagic below)

sev- replied:
It always happens. The comment is related to the unknown logic of weapon choice. Should it be random or the monster could pick the best of its weapons...

bluegr commented on engines/wage/combat.cpp:
```diff
@@ -0,0 +1,915 @@
+		return;
+
+	// TODO: verify that a player not aiming will always target the chest??
+	int targetIndex = -1;
```
I assume that this comment refers to the case where targetIndex is -1, i.e. when weapon->_type is Obj::MAGICAL_OBJECT, and chance < attacker->_physicalAccuracy. Should there be an error message in this case? Or will out of memory access occur?

sev- replied:
It is a todo to clarify the actual engine behavior.

bluegr commented on engines/wage/combat.cpp:
```diff
@@ -0,0 +1,915 @@
+
+	bool usesDecremented = false;
+	int chance = _rnd->getRandomNumber(255);
+	// TODO: what about obj accuracy
```
This comment is a bit cryptic. What does object accuracy refer to?

sev- replied:
Obj::_accuracy variable. An attribute of a weapon object.

bluegr commented on engines/wage/combat.cpp:
```diff
@@ -0,0 +1,915 @@
+	if (targetIndex != -1) {
+		Obj *armor = victim->_armor[targetIndex];
+		if (armor != NULL) {
+			// TODO: Absorb some damage.
```
Test case? Do we need some warning here?

sev- replied:
It always happens, but not implemented as the logic is not known.

bluegr commented on engines/wage/combat.cpp:
```diff
@@ -0,0 +1,915 @@
+	}
+
+	if (causesSpiritualDamage) {
+		/* TODO */
```
Please add a warning here, it'll help us track a test case

sev- replied:
Done.

bluegr commented on engines/wage/combat.cpp:
```diff
@@ -0,0 +1,915 @@
+		}
+
+	char buf[256];
+	snprintf(buf, 256, "%s%s runs %s.", chr->getDefiniteArticle(true), chr->_name.c_str(), directionsS[dir]);
```
Again, we could use Common::String here, together with Common::String::format()

bluegr commented on engines/wage/combat.cpp:
```diff
@@ -0,0 +1,915 @@
+		} else {
+			snprintf(buf, 256, "You now have the %s.", obj->_name.c_str());
+			appendText(buf);
+		}
```
This is a bit redundant, and could be simplified as:

    switch (type) {
    case Chr::HEAD_ARMOR:
    case Chr::BODY_ARMOR:
    case Chr::SHIELD_ARMOR:
    case Chr::MAGIC_ARMOR:
        snprintf(buf, 256, "You are now wearing the %s.", obj->_name.c_str());
        break;
    default:
        snprintf(buf, 256, "You now have the %s.", obj->_name.c_str());
        break;
    }

    appendText(buf);

sev- replied:
Not really. ToDo is to check the actual messages for these armor types.

bluegr commented on engines/wage/design.cpp:
```diff
@@ -0,0 +1,892 @@
+		needRender = true;
+	}
+
+#if 0
```
Why is this bit disabled? Are the dialog graphics not drawn completely?

It seems that the main difference is the line thickness?

sev- replied:
That is for debugging purposes. Will be removed eventually.

bluegr commented on engines/wage/detection.cpp:
```diff
@@ -0,0 +1,240 @@
+	{"afm", "Another Fine Mess"},
+	{"amot", "A Mess O' Trouble"},
+	{"cantitoe", "Camp Cantitoe"},
+	{"scepters", "Enchanced Scepters"},
```
Typo: Enchanted Scepters

sev- replied:
Fixed

bluegr commented on engines/wage/detection.cpp:
```diff
@@ -0,0 +1,240 @@
+	{"cantitoe", "Camp Cantitoe"},
+	{"scepters", "Enchanced Scepters"},
+	{"wage", "WAGE"},
+	{0, 0}
```
Just a note here: AFAIK, there were a lot of games using this engine, about 50+. A reminder that they should be added in the detector as well - these are just a small subset

sev- replied:
Not really. I do not want to pollute the global namespace with minor games. See ADGF_GENERIC define. Rest of the games use 'wage' target.

bluegr commented on engines/wage/gui.cpp:
```diff
@@ -0,0 +1,724 @@
+	}
+
+	drawBox(g, x, y, size, size);
+	drawBox(g, x+width-size-1, y, size, size);
```
Formatting (applies to this and all the drawBox lines below)

sev- replied:
Fixed.

bluegr commented on engines/wage/gui.cpp:
```diff
@@ -0,0 +1,724 @@
+					}
+				}
+			}
+			fillRect(g, x + width - 13, y + size + ARROW_H, 8, height - 2*size - 1 - ARROW_H*2);
```
Formatting

sev- replied:
Fixed.

bluegr commented on engines/wage/gui.cpp:
```diff
@@ -0,0 +1,724 @@
+			for (int yy = 0; yy < ARROW_H; yy++) {
+				for (int xx = 0; xx < ARROW_W; xx++) {
+					if (arrowPixels[yy][xx] != 0) {
+						g->hLine(x1+xx, y1+yy, x1+xx, kColorBlack);
```
Formatting

sev- replied:
Fixed.

bluegr commented on engines/wage/gui.cpp:
```diff
@@ -0,0 +1,724 @@
+	0, 0, 3, 3, 3, 0, 0, 3, 3, 3, 3,
+};
+
+static void cursor_timer_handler(void *refCon) {
```
Use CamelCase here, like in the other parts

sev- replied:
Done.

bluegr commented on engines/wage/gui.cpp:
```diff
@@ -0,0 +1,724 @@
+
+	_inTextSelection = true;
+
+	warning("x: %d y: %d", _selectionStartX, _selectionStartY);
```
I assume that the warnings here are left because this function isn't implemented completely? Please, add a TODO if this is the case

sev- replied:
I am working on it right now. These commits were made this afternoon.

bluegr commented on engines/wage/gui.cpp:
```diff
@@ -0,0 +1,724 @@
+}
+
+void Gui::updateTextSelection(int x, int y) {
+	warning("x: %d y: %d", x, y);
```
Same as startMarking()

sev- replied:
Same reply :)

bluegr commented on engines/wage/combat.cpp:
```diff
@@ -0,0 +1,915 @@
+	Scene *playerScene = _world->_player->_currentScene;
+	const char *msg = playerScene->_messages[dir].c_str();
+
+	warning("Dir: %s  msg: %s", dirName, msg);
```
Leftover debug?

sev- replied:
Right. Killed it.

bluegr commented on engines/wage/script.cpp:
```diff
@@ -0,0 +1,1127 @@
+			Common::String name = lhs->_value.designed->_name;
+			name.toLowercase();
+
+			warning("%s <> %s", name.c_str(), str.c_str());
```
Leftover debug?

sev- replied:
Right

bluegr commented on engines/wage/wage.cpp:
```diff
@@ -0,0 +1,502 @@
+				_world->move(chr, _world->_scenes[key]);
+
+				if (chr->_playerCharacter)
+					warning("Initial scene: %s", key.c_str());
```
Leftover debug?

sev- replied:
Actually it is needed, so I put it under debug()

bluegr commented on engines/wage/world.cpp:
```diff
@@ -0,0 +1,510 @@
+	Common::String *message;
+	if ((message = loadStringFromDITL(resMan, 2910, 1)) != NULL) {
+		message->trim();
+		warning("_gameOverMessage: %s", message->c_str());
```
Leftover debug? (same with the other three warnings below)

sev- replied:
Hid them too

bluegr commented:
Overall, this looks to be in a good state, kudos! :) I've left some comments here and there. Some further testing with the rest of the games would be nice, too (there are over 50 of them)

wjp commented on engines/wage/.gitignore:
```diff
@@ -0,0 +1,3 @@
+.deps
```
This file (.gitignore) shouldn't be here. These things are handled by the main .gitignore.

sev- replied:
Right, it was from the time when the engine lived in a separate directory. Removed.

wjp commented on engines/wage/combat.cpp:
```diff
@@ -0,0 +1,914 @@
+
+static const int directionsX[] = { 0, 0, 1, -1 };
+static const int directionsY[] = { -1, 1, 0, 0 };
+static const char *directionsS[] = { "north", "south", "east", "west" };
```
static const char *const

sev- replied:
fixed

wjp commented on engines/wage/combat.cpp:
```diff
@@ -0,0 +1,914 @@
+
+			Scene *scene = _world->getSceneAt(destX, destY);
+
+			if (scene != NULL && scene->_chrs.size() == 0) {
```
_chrs.empty(), since that's a Common::List

wjp commented on engines/wage/combat.cpp:
```diff
@@ -0,0 +1,914 @@
+}
+
+void WageEngine::takeObj(Obj *obj) {
+  if ((int)_world->_player->_inventory.size() >= _world->_player->_maximumCarriedObjects) {
```
(Minor: wrong indent)

wjp commented on engines/wage/combat.cpp:
```diff
@@ -0,0 +1,914 @@
+
+bool WageEngine::handleMoveCommand(Directions dir, const char *dirName) {
+	Scene *playerScene = _world->_player->_currentScene;
+	const char *msg = playerScene->_messages[dir].c_str();
```
Might be cleaner to move the c_str() down to the appendText calls to avoid the strlen, but probably unimportant.

wjp commented on engines/wage/combat.cpp:
```diff
@@ -0,0 +1,914 @@
+			objs.push_back(*it);
+
+	if (!objs.size()) {
+		appendText((char *)"Your pack is empty.");
```
Unnecessary cast.

wjp replied:
(There are a few more of these.)

sev- replied:
Yep. I changed the method footprint, and forgot about these. Fixed.

wjp commented on engines/wage/combat.cpp:
```diff
@@ -0,0 +1,914 @@
+	return true;
+}
+
+static const char *armorMessages[] = {
```
static const char *const

wjp commented on engines/wage/design.cpp:
```diff
@@ -0,0 +1,534 @@
+			_visited[y * _w + x] = 1;
+			*p = _color2;
+
+			Common::Point *pt = new Common::Point(x, y);
```
I don't think using a List of Point pointers makes sense here compared to simply using a List of Point objects. (The object is only 4 bytes.)

lordhoto replied:
Yup, same thoughts here :-).

wjp commented on engines/wage/design.h:
```diff
@@ -0,0 +1,115 @@
+		return _bounds;
+	}
+
+    void paint(Graphics::Surface *canvas, Patterns &patterns, int x, int y);
```
Minor: there are a few wrong indents in this header)

sev- replied:
Fixed.

wjp commented on engines/wage/detection.cpp:
```diff
@@ -0,0 +1,240 @@
+	Common::StringArray filenames;
+	char saveDesc[31];
+	Common::String pattern = target;
+	pattern += ".???";
```
You can use @lordhoto's new ".###' pattern here, probably.

sev- replied:
Where is that? I do not see '#' handling in String::matchString()

wjp replied:
06641f29a7bdcda280b0291f215193f055c83969

I suppose it was added after you branched for wage, so you'll probably have to merge in a recent master for this.

wjp commented on engines/wage/design.cpp:
```diff
@@ -0,0 +1,534 @@
+	if (p->thickness == 1) {
+		if (x >= 0 && x < p->surface->w && y >= 0 && y < p->surface->h) {
+			*((byte *)p->surface->getBasePtr(x, y)) =
+				(pat[y % 8] & (1 << (7 - x % 8))) ?
```
These %8's will produce suboptimal code if optimizations are disabled since x and y are signed.

sev- replied:
Right, copied it to units

wjp commented on engines/wage/entities.cpp:
```diff
@@ -0,0 +1,500 @@
+
+// Source: Apple IIGS Technical Note #41, "Font Family Numbers"
+// http://apple2.boldt.ca/?page=til/tn.iigs.041
+const char *fontNames[] = {
```
const char *const (and maybe static)

sev- replied:
Fixed

wjp commented on engines/wage/gui.cpp:
```diff
@@ -0,0 +1,856 @@
+	0, 0, 3, 3, 3, 0, 0, 3, 3, 3, 3,
+};
+
+static void cursorTimerHandler(void *refCon) {
```
This function scares me a bit. It seems to have some race conditions since setting cursoroff/cursorstate isn't protected by a mutex, and I'm not convinced it's safe to call copyRectToScreen from the timer thread.

And, should it always draw the cursor, even if _cursorOff?

sev- replied:
Heh, I had a race condition there :). See Gui::drawInput()

wjp replied:
Yeah, I saw that call in there. I don't really understand how that makes the logic thread-safe, though.

Also, I asked @lordhoto about calling copyRectToScreen from a thread, and he pointed out the comment around system.h line 400, that says calls to OSystem graphics functions aren't guaranteed to be thread-safe.

lordhoto replied:
In general I would suggest to not call *any* graphics functionality from a timer.
`updateScreen` and `setMouseCursor` are *completely* unsafe to call from a timer for OpenGL based outputs. This is luckily not done here.
The other ones aren't thread-safe, but usually work with OpenGL output nowadays due to delayed updates.
I think we should update our `OSystem` documentation to reflect this.

sev- replied:
I rewrote it for drawing in the main thread.

wjp commented on engines/wage/menu.cpp:
```diff
@@ -0,0 +1,563 @@
+	int action;
+	byte shortcut;
+	bool enabled;
+} menuSubItems[] = {
```
static const?

sev- replied:
Right

wjp commented on engines/wage/script.h:
```diff
@@ -0,0 +1,178 @@
+		Common::String _str;
+
+		Operand(Obj *value, OperandType type) {
+			_value.obj = value;
```
Would asserts that the type matches be appropriate here?

sev- replied:
What do you mean? That is overloaded constructor, and there are no type casts in the callers.

wjp replied:
I was thinking of (presumably invalid?) calls like

     new Operand(Common::String("x"), INT);

lordhoto replied:
It might also make sense to remove  the `type` parameter when the `type` can be uniquely determined by the `value` parameter's type. (And then make those single parameter constructors `explicit` to prevent any silent conversions.)

sev- replied:
Well, I control what I do with Operands, and that piece of code is feature complete. I'd rather not touch it.

wjp commented on engines/wage/script.cpp:
```diff
@@ -0,0 +1,1126 @@
+struct Mapping {
+	const char *cmd;
+	int type;
+} mapping[] = {
```
static const?

wjp commented on engines/wage/util.cpp:
```diff
@@ -0,0 +1,126 @@
+
+	len = in->readSByte();
+	if (len < 0)
+		len += 256;
```
Isn't this just readByte(), then?

sev- replied:
I'm not sure, need to test.

lordhoto replied:
Yup, this code has the same result as doing `readByte()` in the first place.

sev- replied:
Fixed and tested.

wjp commented on engines/wage/menu.cpp:
```diff
@@ -0,0 +1,563 @@
+			bool enabled = true;
+			int style = 0;
+			char shortcut = 0;
+			char *shortptr = strrchr(item.c_str(), '/');
```
This fails to compile because strrchr expects a char* (not const char*).

sev- replied:
Which platform? I have the following prototype:
char *strrchr(const char *s, int c);

wjp replied:
Ah, in fact I seem to have two:

    char *strchr (char *__s, int __c)
    const char *strchr (const char *__s, int __c)

(and same for strrchr). This is glibc 2.21

lordhoto replied:
We need to use `const char *` for `shortptr` here. Otherwise the compiler will try to pick the version of `strrchr` taking `char *` resulting in a compilation error.

Minor side remark: this should probably be named `shortPtr` according to our guidelines.

sev- replied:
Done and renamed

wjp commented on engines/wage/world.cpp:
```diff
@@ -0,0 +1,510 @@
+		sprintf(menu, "About %s...", _name.c_str());
+	} else { // Replace '@' with name
+		const char *str = _aboutMenuItemName.c_str();
+		char *pos = strchr(str, '@');
```
This also doesn't compile due to strchr expecting a char*.

lordhoto replied:
We need to use `const char *` for `pos` here. Otherwise the compiler will try to pick the version of `strchr` taking `char *` resulting in a compilation error.

sev- replied:
Fixed

wjp commented:
Overall this looks very clean; nice work.

Only two real issues for me: I added line comments to two places that make it fail to build for me, and I don't know if the cursor handling in a timer proc is safe/allowed.

sev- commented:
Guys, I'd like to merge this and continue in-tree.

dreammaster commented:
For what it's worth from me, there's been adequate time, discussion, and
fixes already. And the engine is in a good enough state for merging in from
what little I've looked at it. So I'd say go for it.

On Sat, Feb 13, 2016 at 5:44 PM, Eugene Sandulenko <notifications@github.com
> wrote:

> Guys, I'd like to merge this and continue in-tree.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scummvm/scummvm/pull/658#issuecomment-183767044>.
>

lordhoto commented on engines/wage/combat.cpp:
```diff
@@ -0,0 +1,914 @@
+
+	int token = hat.drawToken();
+	switch (token) {
+		case kTokWeapons:
```
(Minor) We don't increase indent level for cases.

sev- replied:
Fixed.

lordhoto commented on engines/wage/combat.cpp:
```diff
@@ -0,0 +1,914 @@
+	delete magics;
+}
+
+const char *targets[] = { "head", "chest", "side" };
```
Looks like these are never modified, thus `static const char *const` seems more appropriate.

sev- replied:
Fixed.

lordhoto commented on engines/wage/combat.cpp:
```diff
@@ -0,0 +1,914 @@
+	return true;
+}
+
+Common::String *WageEngine::getGroundItemsList(Scene *scene) {
```
It looks like this is used in two places. One right above, which checks for `NULL` as possible return value, and one in `WageEngine::attackHit` which does not care about it possibly returning `NULL`. Since it looks like when this returns non-`NULL` is always empty it might be more sensible to return a `Common::String` and perform `empty` checks. That also avoids manual memory management.

lordhoto commented on engines/wage/combat.cpp:
```diff
@@ -0,0 +1,914 @@
+	return NULL;
+}
+
+void WageEngine::appendObjNames(Common::String &str, ObjArray &objs) {
```
This doesn't seem to modify contents of `objs`, thus I suggest making it `const`. I noticed `getIndefiniteArticle`, takes a non-`const` reference to `Common::String`, but since it doesn't change its contents either, it can easily be modified to take a `const` reference too, allowing this method (and maybe others) to use `const` in more places too.

sev- replied:
Fixed in both cases.

lordhoto commented on engines/wage/combat.cpp:
```diff
@@ -0,0 +1,914 @@
+		if ((*it)->_type != Obj::IMMOBILE_OBJECT)
+			objs.push_back(*it);
+
+	if (objs.size()) {
```
(Minor) I would suggest to always use `!container.empty()` instead of `container.size()` that way you have a consistent check for _is not empty_ which also doesn't do unnecessary computation for types like lists.

sev- replied:
Sigh. Adjusted everywhere.

lordhoto commented on engines/wage/combat.cpp:
```diff
@@ -0,0 +1,914 @@
+	return false;
+}
+
+bool WageEngine::tryAttack(Obj *weapon, Common::String &input) {
```
(Minor) This can also take `const` versions of both parameters. Might even make sense to make `weapon` a reference, but as far as I understand it, the only caller has a pointer anyway, so this isn't too important. But in general using (const) references can clarify that the parameter should always be non-`NULL`.

sev- replied:
Fixed.

lordhoto commented on engines/wage/design.cpp:
```diff
@@ -0,0 +1,539 @@
+
+namespace Wage {
+
+struct plotData {
```
(Minor) `PlotData` according to our naming guidelines.

sev- replied:
Fixed

lordhoto commented on engines/wage/design.cpp:
```diff
@@ -0,0 +1,539 @@
+
+Design::~Design() {
+	free(_data);
+	if (_surface && _surface->getPixels())
```
(Minor) `Surface::free` also works when `getPixels` returns `NULL`, thus the second part of the check is not necessary.

sev- replied:
Fixed

lordhoto commented on engines/wage/design.cpp:
```diff
@@ -0,0 +1,539 @@
+		int y2 = y1;
+		int x2 = x1;
+		int b = in.readSByte();
+		if ((b & 0xff) == 0x80) {
```
(Minor) This can be simplified to `b == -128` (same applies to below).

sev- replied:
Fixed

lordhoto commented on engines/wage/design.cpp:
```diff
@@ -0,0 +1,539 @@
+
+	int npoints = xcoords.size();
+	int *xpoints = (int *)calloc(npoints, sizeof(int));
+	int *ypoints = (int *)calloc(npoints, sizeof(int));
```
It looks like `xpoints` and `ypoints` are never modified during their lifetime. Thus, I think we can easily remove their creation and use `xcoords` and `ycoords` contents directly.

sev- replied:
In this case I will not be able to pass it to Graphics::drawPolygonScan()

lordhoto replied:
You can easily do `xcoords.begin()` and `ycoords.begin()` to pass them to `Graphics::drawPolygonScan`

lordhoto commented on engines/wage/design.cpp:
```diff
@@ -0,0 +1,539 @@
+	}
+
+	while (numBytes--)
+		in.readByte();
```
It looks like `drawBitmap` is always used on a `SeekableReadStream`, this would allow to make `in` of such type and thus allow us to use `skip`/`seek` here.

sev- replied:
Fixed

lordhoto commented on engines/wage/design.cpp:
```diff
@@ -0,0 +1,539 @@
+
+struct plotData {
+	Graphics::Surface *surface;
+	Patterns *patterns;
```
To me it looks we can make this `const Patterns &` and simplify a lot of code, i.e. use `[]` directly etc.

lordhoto commented on engines/wage/detection.cpp:
```diff
@@ -0,0 +1,240 @@
+
+void WageMetaEngine::removeSaveState(const char *target, int slot) const {
+	char fileName[MAXPATHLEN];
+	sprintf(fileName, "%s.%03d", target, slot);
```
I suggest to use `Common::String::format` and let the savefile manager worry about length.

sev- replied:
Fixed.

lordhoto commented on engines/wage/entities.h:
```diff
@@ -0,0 +1,345 @@
+
+class Sound {
+public:
+  Sound(Common::String name, Common::SeekableReadStream *data) : _name(name), _data(data) {}
```
(minor) this doesn't use tab for indentation

sev- replied:
Fixed

lordhoto commented on engines/wage/entities.h:
```diff
@@ -0,0 +1,345 @@
+class Sound {
+public:
+  Sound(Common::String name, Common::SeekableReadStream *data) : _name(name), _data(data) {}
+	~Sound() { }
```
(minor) superflous explicit destructor

sev- replied:
This is a TODO. Sound is not yet implemented

lordhoto commented on engines/wage/randomhat.cpp:
```diff
@@ -0,0 +1,83 @@
+namespace Wage {
+
+void RandomHat::addTokens(int type, int count) {
+	_tokens.setVal(type, _tokens.getVal(type, 0) + count);
```
So, here I was to suggest we can simplify this to `_tokens[type] += count;` because we default initialize values. And then I saw that we don't on PS2 because of internal compiler errors happening when we do.... _sigh_

sev- replied:
Sigh

lordhoto commented on engines/wage/script.h:
```diff
@@ -0,0 +1,178 @@
+				delete _value.string;
+		}
+
+		Common::String toString() {
```
I sould suggest to move the implementation to `script.cpp`.

sev- replied:
Moved.

lordhoto commented on engines/wage/script.h:
```diff
@@ -0,0 +1,178 @@
+		Common::String toString() {
+			char buf[128];
+
+			if (_value.obj == NULL)
```
This check looks odd, what is the purpose? It looks like none of the cases below would check for a `NULL` pointer and thus crash anyway?

sev- replied:
This is a leftover from Java code. Removed

lordhoto commented on engines/wage/script.h:
```diff
@@ -0,0 +1,178 @@
+			Designed *inputClick;
+		} _value;
+		OperandType _type;
+		Common::String _str;
```
I don't get the purpose of this member, it looks like it's only used locally in `toString`. Since `toString` doesn't return a const reference, there doesn't seem to be any need to have this around though. Thus I think we can simply remove it.

sev- replied:
This is a remnant from the original Java code. Removed

lordhoto commented on engines/wage/script.h:
```diff
@@ -0,0 +1,178 @@
+			switch(_type) {
+			case NUMBER:
+				_str = snprintf(buf, 128, "%d", _value.number);
+				return _str;
```
These two lines might be simplified to `return Common::String::format("%d", _value.number);`

sev- replied:
Fixed

lordhoto commented on engines/wage/util.cpp:
```diff
@@ -0,0 +1,126 @@
+}
+
+enum {
+	GENDER_HE = 0,
```
(Minor) since this talks about gender it's more natural to use male, female, neuter here.

sev- replied:
Renamed, thanks for suggestion

lordhoto commented on engines/wage/wage.cpp:
```diff
@@ -0,0 +1,505 @@
+	_gui->appendText(str);
+
+	if (_inputText.size())
+		_inputText = "";
```
This can be simplified to `_inputText.clear()` and the `size` (again really a `!empty` check) seems superflous.

sev- replied:
Fixed

lordhoto commented on engines/wage/wage.h:
```diff
@@ -0,0 +1,234 @@
+ */
+
+#ifndef WAGE_H
+#define WAGE_H
```
`WAGE_WAGE_H`

sev- replied:
Fixed

lordhoto commented on engines/wage/world.cpp:
```diff
@@ -0,0 +1,510 @@
+	return NULL;
+}
+
+bool InvComparator(Obj *l, Obj *r) {
```
This is a function thus `invComparator` according to our guidelines. It can be made `static` too.

sev- replied:
Renamed

lordhoto commented on engines/wage/world.cpp:
```diff
@@ -0,0 +1,510 @@
+	_engine->onMove(obj, from, chr);
+}
+
+bool ObjComparator(Obj *o1, Obj *o2) {
```
Same as for `InvComparator`. Parameters can be made `const` too (which I forgot to mention earlier).

sev- replied:
Done both.

lordhoto commented on engines/wage/world.cpp:
```diff
@@ -0,0 +1,510 @@
+	_engine->onMove(obj, from, scene);
+}
+
+bool ChrComparator(Chr *l, Chr *r) {
```
Same as for `ObjComparator`.

sev- replied:
Done

lordhoto commented on graphics/primitives.cpp:
```diff
@@ -79,4 +96,344 @@ void drawThickLine(int x0, int y0, int x1, int y1, int penX, int penY, int color
+			j = i;
+		}
+
+		//  Sort the nodes, via a simple “Bubble” sort.
```
Not sure if it's worth it but one could simply use `Common::sort` here.

sev- replied:
Right, fixed.

lordhoto commented on graphics/primitives.cpp:
```diff
@@ -79,4 +96,344 @@ void drawThickLine(int x0, int y0, int x1, int y1, int penX, int penY, int color
+
+		for (i = 0; i < npoints; i++) {
+			if ((polyY[i] < pixelY && polyY[j] >= pixelY) || (polyY[j] < pixelY && polyY[i] >= pixelY)) {
+				nodeX[nodes++] = (int)(polyX[i] + (double)(pixelY - polyY[i]) / (double)(polyY[j]-polyY[i]) *
```
We might want to replace the floating point math here.

sev- replied:
I was getting nasty rounding errors with integers and even floats.

lordhoto commented on engines/wage/gui-console.cpp:
```diff
@@ -0,0 +1,398 @@
+
+	// Okay, we got new lines, need to split it
+	// and push substrings individually
+	Common::String tmp = "";
```
There's no need to initialize it with `""`.

sev- replied:
Fixed

lordhoto commented on engines/wage/gui-console.cpp:
```diff
@@ -0,0 +1,398 @@
+		if (str[i] == '\n') {
+			_out.push_back(tmp);
+			flowText(tmp);
+			tmp = "";
```
`tmp.clear()`

sev- replied:
Done

lordhoto commented on engines/wage/menu.cpp:
```diff
@@ -0,0 +1,563 @@
+			menu->subitems.push_back(new MenuSubItem(item.c_str(), kMenuActionCommand, style, shortcut, enabled));
+		}
+
+		item = "";
```
`item.clear()`

sev- replied:
Done

lordhoto commented on engines/wage/wage.cpp:
```diff
@@ -0,0 +1,505 @@
+					break;
+
+				processTurn(&_inputText, NULL);
+				_inputText = "";
```
`_inputText.clear()`

sev- replied:
Done

lordhoto commented on engines/wage/world.cpp:
```diff
@@ -0,0 +1,510 @@
+	res = resMan->getResource(MKTAG('M','E','N','U'), 2001);
+	if (res != NULL) {
+		Common::StringArray *menu = readMenu(res);
+		_aboutMenuItemName = "";
```
`_aboutMenuItemName.clear()`

sev- replied:
Done

lordhoto commented:
I think overall it looks in a decent state. Good work. We'll need to fix the compilation errors before merge though. We might even want to fixup history here to help bisection in the future. @wjp any thoughts here?

I agree with @bluegr that it would be nice to have more comments. But we don't really have that in many other engines either, so I don't really see this as a hard requirement.

Some random thoughts which would be nice to see resolved later:
* There's a lot of pointer uses which seem unnecessary. Avoiding pointers when unecessary but using either values or const references instead could simplify a lot of code.
* Adding `const` to more places would definitely be nice too.
* I think we should strive for full names instead of abbreviations to ease readability. For example, `Chr` could be made `Character`.
* A lot of places use `size()` of containers for _is empty_ or _is not empty_ checks. I really suggest using `empty()` which conveys the intention of what you are trying to do much better and also has the benefit of being O(1) for our containers.
* There're a lot of warnings about signed/unsigned comparison in loops using `int` for their iterator and comparing it to `container.size()`

lordhoto commented on engines/wage/entities.cpp:
```diff
@@ -0,0 +1,500 @@
+	_context._statVariables[PHYS_SPE_BAS] = _context._statVariables[PHYS_SPE_CUR] = _runningSpeed;
+}
+
+ObjArray *Chr::getWeapons(bool includeMagic) {
```
We can simply return an `ObjArray` instead of a pointer here. This will simplify caller code by a lot.
RVO will likely be applied and avoid an unnecessary copy. And even if not, the copy should be reasonably small to not cause any issues (and if it does, we can worry about it later).

lordhoto commented on engines/wage/entities.cpp:
```diff
@@ -0,0 +1,500 @@
+	return list;
+}
+
+ObjArray *Chr::getMagicalObjects() {
```
Same as for `Chr::getWeapons` a few lines above.

sev- commented:
> We'll need to fix the compilation errors before merge though.

Seems to be fixed

> We might even want to fixup history here to help bisection in the future. @wjp any thoughts here?

I am not sure what you're referring to here

> I agree with @bluegr that it would be nice to have more comments. But we don't really have that in many other engines either, so I don't really see this as a hard requirement.

There was no goal to have the engine absolutely clear to everyone (as in every engine we have). Moreover, some of the confusion was related to the engine internals, not to the code itself

> There's a lot of pointer uses which seem unnecessary. Avoiding pointers when unecessary but using either values or const references instead could simplify a lot of code.

I addressed few cases which you referred.

> Adding const to more places would definitely be nice too.

That was addressed

> I think we should strive for full names instead of abbreviations to ease readability. For example, Chr could be made Character.

Wouldn't that just bloat the code?

> A lot of places use size() of containers for is empty or is not empty checks. I really suggest using empty() which conveys the intention of what you are trying to do much better and also has the benefit of being O(1) for our containers.

Fixed everywhere (I hope)

> There're a lot of warnings about signed/unsigned comparison in loops using int for their iterator and comparing it to container.size()

I am not getting those on Mac, will try on Linux which is less forgiving. But in any case I expect more errors/warnings to be uncovered once the buildbot deals with it.

sev- commented:
Fixed int/uint mess

sev- commented:
Thanks for the review, guys, I'll merge this in an hour or so, once I have enough time to fix buildbot if needed.

lordhoto commented:
>> We might even want to fixup history here to help bisection in the future. @wjp any thoughts here?
> I am not sure what you're referring to here

To the compilation errors due to `strrchr` and `strchr`. If we fix the when they were introduced by rewriting the history, we can easily do bisections over all intermediate revisions.

> Wouldn't that just bloat the code?

I don't see why it would.

sev- commented:
Fixed history. Merging now.
